tools/halide_benchmark.h

```diff
@@ -20,19 +59,69 @@ namespace Tools {
 // for real-world use. For now, callers using this to benchmark GPU
 // code should measure with extreme caution.
 
-template <typename F>
-double benchmark(int samples, int iterations, F op) {
-    double best = std::numeric_limits<double>::infinity();
-    for (int i = 0; i < samples; i++) {
-        auto t1 = std::chrono::high_resolution_clock::now();
-        for (int j = 0; j < iterations; j++) {
-            op();
+struct BenchmarkConfig {
+    // Terminate when the relative difference between the best runtime
+    // seen and the third-best runtime seen is no more than
+    // this. Controls accuracy. The closer to zero this gets the more
+    // reliable the answer, but the longer it may take to run.
+    double accuracy {0.03};
+    
+    // Take at least this much time per sample (seconds). Prevents
+    // benchmarking overhead and clock resolution being a major factor
+    // for very cheap benchmarks.
+    double min_time {0.005};
+
+    // Run at least this many iterations per sample.
+    uint64_t min_iters {1};
+
+    // Run at most this many iterations over all samples.
+    uint64_t max_iters {1000};
+};
+
+struct BenchmarkResult {
+    // Average elapsed wall-clock time per iteration (seconds).
+    double wall_time;
+
+    // Actual number of iterations per sample.
+    uint64_t iterations;
+
+    // Number of samples taken.
+    uint64_t samples;
+
+    operator double() const {return wall_time;}
+};
+
+template<typename Op>
+BenchmarkResult benchmark(Op op,
+                          const BenchmarkConfig& config = BenchmarkConfig()) {
+    BenchmarkResult result {0, 0, 0};
+
+    uint64_t total_iters = 0;
+    double times[] = {0, 0, 0, 0};
+    for (uint64_t iters_per_sample = config.min_iters;
+         total_iters < config.max_iters;
+         iters_per_sample *= 2) {
+        times[0] = benchmark(1, iters_per_sample, op);
```

Comment:
One drawback to this approach is that Really Long Benchmarks now have to run at least 4 times; the GTest-ish approach means that we bail out after a single run.

Reply:
Three times usually - it's possible to not enter the loop that adds additional samples if the first three times are similar. This guards against fast tests run during a system hiccup that makes them take unusually long one time. On average this seems to speed things up (~81s in test_performance in this branch on the buildbots, vs ~160s in the other branch, vs ~100s in master)